lib/process_settings/monitor.rb

```diff
@@ -63,15 +66,35 @@ def stop
       @listener&.stop
     end
 
-    # TODO:
-    # 1. rename this to `when_updated` and clone that interface from InvocaCluster
-    # 2. since the callback yields self, support Instance#[]; have ProcessSettings.[] delegate there.
+    # Idempotently adds the given block to the when_updated collection
+    # calls the block first unless initial_update: false is passed
+    # returns a handle (the block itself) which can later be passed into cancel_when_updated
+    def when_updated(initial_update: true, &block)
+      if @when_updated_blocks.add?(block)
+        if initial_update
+          begin
+            block.call(self)
+          rescue => ex
+            logger.error("ProcessSettings::Monitor#when_updated rescued exception during initialization:\n#{ex.class}: #{ex.message}")
+          end
+        end
+      end
+
+      block
+    end
+
+    # removes the given when_updated block identified by the handle returned from when_updated
+    def cancel_when_updated(handle)
+      @when_updated_blocks.delete_if { |callback| callback.eql?(handle) }
+    end
 
     # Registers the given callback block to be called when settings change.
     # These are run using the shared thread that monitors for changes so be courteous and don't monopolize it!
+    # @deprecated
     def on_change(&callback)
       @on_change_callbacks << callback
```

Comment:
I was thinking we'd store the `on_change` callbacks onto the `@when_updated_blocks` collection starting now. It would be a bit DRYer that way. But then again, the way you have keeps the two versions of the feature fully separated from each other.

Reply:
I thought about doing that. I think while this way is deprecated, it's better and cleaner to keep them completely separated to keep the logic untangled. Makes removal of the feature in `1.0.0` simpler, and also makes sure that no development done between `0.7.0` and `1.0.0` accidentally changes the functionality of the deprecated method.

Reply:
👍